# SVM_SVR

- SVM Kernel & ROC AUC Curve Explanation 

- 1st we can see the "SkLearn" libary for the SVC Support vector classifier.
- For this Google it -> Sklearn SVC.
- In this document we can see -> URL : [ https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html ]
- Here we can SVC libary has been given -> This class is already Embedded ->
- If we open the dicument we can see this URL Class :
- class sklearn.svm.SVC(*, C=1.0, kernel='rbf', degree=3, gamma='scale', coef0=0.0, shrinking=True, probability=False, tol=0.001, cache_size=200, class_weight=None, verbose=False, max_iter=-1, decision_function_shape='ovr', break_ties=False, random_state=None)[source]

- & they have given different different Parameters has given.
- [ *, C=1.0, kernel='rbf', degree=3, gamma='scale', coef0=0.0, shrinking=True, probability=False, tol=0.001, cache_size=200, class_weight=None, verbose=False, max_iter=-1, decision_function_shape='ovr', break_ties=False, random_state=None ]

- So,this parameters is nothing whatever Mathematics we have learn, for those formula and all we are using, this Hyperparaameter.
- So, we will build simple Classifier over here. So this URL is SVC - "Support verctor classifier".
  
- Just Open This URL : ["https://scikit-learn.org/stable/api/sklearn.svm.html"]
- over here, we can see the : "Estimators" a different diffferent implementation has been given 
- For Example :
- 1.LinearSVC : Linear Support Vector Classification.
- 2.LinearSVR : Linear Support Vector Regression.
- 3.NuSVC : Nu-Support Vector Classification.
- 4.OneClassSVM : Unsupervised Outlier Detection.
- 5.SVC : C-Support Vector Classification. URL : [ https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html#sklearn.svm.SVC ]
- 6.SVR : Epsilon-Support Vector Regression. URL : [ https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html#sklearn.svm.SVR]
- 7.l1_min_c : Return the lowest bound for C.

- So, simply we are going to do a SVC and SVR
- So lets open the
- 5.SVC UrL : C-Support Vector Classification. URL : https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html#sklearn.svm.SVC
- 6.SVR : Epsilon-Support Vector Regression. URL : https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html#sklearn.svm.SVR

- So, lets take one dataset and we can import over here.
- We will going to take the Wine Quality dataset, we will create SVC classifier on top that dataset.
- This is my Raw data URL : [ "https://raw.githubusercontent.com/aniruddhachoudhury/Red-Wine-Quality/master/winequality-red.csv" ]
- This dataset, we can read over here - URL : 

In [1]:
import pandas as pd

In [2]:
pd.read_csv("https://raw.githubusercontent.com/aniruddhachoudhury/Red-Wine-Quality/master/winequality-red.csv")

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.700,0.00,1.9,0.076,11.0,34.0,0.99780,3.51,0.56,9.4,5
1,7.8,0.880,0.00,2.6,0.098,25.0,67.0,0.99680,3.20,0.68,9.8,5
2,7.8,0.760,0.04,2.3,0.092,15.0,54.0,0.99700,3.26,0.65,9.8,5
3,11.2,0.280,0.56,1.9,0.075,17.0,60.0,0.99800,3.16,0.58,9.8,6
4,7.4,0.700,0.00,1.9,0.076,11.0,34.0,0.99780,3.51,0.56,9.4,5
...,...,...,...,...,...,...,...,...,...,...,...,...
1594,6.2,0.600,0.08,2.0,0.090,32.0,44.0,0.99490,3.45,0.58,10.5,5
1595,5.9,0.550,0.10,2.2,0.062,39.0,51.0,0.99512,3.52,0.76,11.2,6
1596,6.3,0.510,0.13,2.3,0.076,29.0,40.0,0.99574,3.42,0.75,11.0,6
1597,5.9,0.645,0.12,2.0,0.075,32.0,44.0,0.99547,3.57,0.71,10.2,5


# Steps : Machine Learning Project 

- 1st we will going to import the data, this is call "Data ingestion". This "Data ingestion" we will do on a Large scale when we do end to end project

- 2nd we will do Exploratory Data Analysis (EDA) 
- 3rd we have to do Pre-processing of the Data.
- 4th we have to create a model : 
- We will going to create now is a Support Vector Machine (SVM), inside there is SVC (Support Vector Classifier), which means if we are not going to 
- transformed our data, in that case, we will going to use a simple, means if we are able to classifier our data, with the help of simple line or in 
- a simple hyper plane in that case we can use the SVC (Support Vector Classifier).
- But lets say, if we going to transform our data from the 1D to 2D, or 2D to 3D in that case, we will going to use this - Support Vector Machine (SVM)
- There is one simlple Formula, because of which we are able to Transform our data.
- "Model" is nothing but Mathematical equation.
- 5th Now we have to evaluate our Model, If I have to Evaluation our model in in Classification promblem statement.

# There are few Tehcnique :

- (a) Accuracy Score 
- (b) Performance Matrics is a nothing Confusion Matrics only. Either of them any one I can choose.
- Insdide this ("Confusion Matrics") we have alot of terms, we can calculate Accuracy and We can calculate the Error. 
- We can Precision vs. Recall vs. F1-Score

- (c) AUC & ROC | AUC (Area Under the Curve) & ROC (Receiver Operating Characteristic)

- Similar way we try to implement the problem statement.

# Difference between SVC and SVR? when and what time we use these 2 options?

- SVC means Classification, we are going to solve the Classifier problem statement
- SVR means Regression, we are going to solve the Regression problem statement
- So, we just Load the "Wine Dataset" this is which classification?
- We need to predict the Quality of the Wine. My dataset we just load is "Classification" Problem statement. We will do some analysis on this load
- Data, then we will get the know more in details.


# Lets keep this data inside the Variable : 

In [3]:
data = pd.read_csv("https://raw.githubusercontent.com/aniruddhachoudhury/Red-Wine-Quality/master/winequality-red.csv")

In [4]:
data

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.700,0.00,1.9,0.076,11.0,34.0,0.99780,3.51,0.56,9.4,5
1,7.8,0.880,0.00,2.6,0.098,25.0,67.0,0.99680,3.20,0.68,9.8,5
2,7.8,0.760,0.04,2.3,0.092,15.0,54.0,0.99700,3.26,0.65,9.8,5
3,11.2,0.280,0.56,1.9,0.075,17.0,60.0,0.99800,3.16,0.58,9.8,6
4,7.4,0.700,0.00,1.9,0.076,11.0,34.0,0.99780,3.51,0.56,9.4,5
...,...,...,...,...,...,...,...,...,...,...,...,...
1594,6.2,0.600,0.08,2.0,0.090,32.0,44.0,0.99490,3.45,0.58,10.5,5
1595,5.9,0.550,0.10,2.2,0.062,39.0,51.0,0.99512,3.52,0.76,11.2,6
1596,6.3,0.510,0.13,2.3,0.076,29.0,40.0,0.99574,3.42,0.75,11.0,6
1597,5.9,0.645,0.12,2.0,0.075,32.0,44.0,0.99547,3.57,0.71,10.2,5


- Lets check what are the Columns are aviable over here inside my dataset.

- As we can lets do with 3 different ways - 
- (a) Head
- (b) Tail
- (c) Quality
- (d) Columns
- (e) Info
- (f) describe.T - Here we can check "Mean" "Median" & "Mode"
    
- So, with this many can we find some conclusion our to this dataset.
- For example : "fixed acidity', 'volatile acidity', 'citric acid' these many features or columns has been given, with respect to the Wine.
- what I need to find out is that, We need to predict the Quality of the Wine.

- So, here in the Row section, or we can we see we are having the Numerical features, all of these are numerical features accept this "Quality" Column.
- What I need to do, base on this features, I am going to predict is the "Quality of the Wine".


# Note :

- This comes under "Supervised Machine Learning", here we are solving the Classifier problem statement.

In [5]:
data.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


In [6]:
data.tail()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
1594,6.2,0.600,0.08,2.0,0.090,32.0,44.0,0.99490,3.45,0.58,10.5,5
1595,5.9,0.550,0.10,2.2,0.062,39.0,51.0,0.99512,3.52,0.76,11.2,6
1596,6.3,0.510,0.13,2.3,0.076,29.0,40.0,0.99574,3.42,0.75,11.0,6
1597,5.9,0.645,0.12,2.0,0.075,32.0,44.0,0.99547,3.57,0.71,10.2,5
1598,6.0,0.310,0.47,3.6,0.067,18.0,42.0,0.99549,3.39,0.66,11.0,6


- Here we can see the "Quality" Column & 
- Here we can also write unique fuction over here. So as we can see we are able to find out the Unique class also.
- how many Classes are there and with respect to this, As we can see total = [5, 6, 7, 4, 8, 3] a Total = 6 classification is there.

# This is a "Multi-class Classification" or "Binary Classification" ?
- This below is the "Multi-class Classification" problem.

# Note : "Binary Classification" :

- The "Binary Classification" we already did previously class like - 
- (a) "algerian forest dataset" (b) "diabetes dataset"


- As of now, I toook this ""Multi-class Classification", the approach will be different. In this "Multi-class Classification", it will be taking
- 1 vs rest and 1 vs 1. We will talk more about it as of now, As of now this is "Classification problem" statement we are going to solve it.

In [7]:
data.quality

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

In [8]:
data.quality.unique()

array([5, 6, 7, 4, 8, 3])

# Note :
- Now if I want print the Values counts, how many count is there. so, we are able to find, different different value count.

In [9]:
data["quality"].value_counts()

quality
5    681
6    638
7    199
4     53
8     18
3     10
Name: count, dtype: int64

In [10]:
data.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

- Here, is there any "Null" Value over here, no right, There is "Non-Null" values over here, So should I handle the Null Value over here.
- No we are not required to handle the "Null Values" over here.

data.info()

- Standara scala, Scalling of that particular data.

- So, we have so many technique interms of EDA and pre-Processing.

- So Why "StandardScaler()" important over here, with respect to problem statement. By seeing this value why "Standard_Scaler" is important because
- different unit.
- The data unit might be a different & Scale might be a very very large.
- How we can check? over here we can check with respect to a Column, we can plot differen different plot also.
- We can check with distribution plot also, we can plot the contile, a Box plot also we can plot, diffrent different plot we are having we can plot 
- ans check, we can check data varaition and all.
- As of now I am going to check with this particular statistic.

- Over here we will be able to find the "Min and Max" Value . 
- This is the Minumum Value : 4.60000 & This is the Maximum value : 15.90000.

- Here this is the "variables" : fixed acidity| volatile acidity | citric acid | residual sugar| chlorides | free sulfur dioxide etc...
- Here we can see that few of the variables Maximum value is very very high which means scale is different.
- Minimum value and Maximum value is very very different.

- Lets check this vairable : ("total sulfur dioxide") inside this variable we will find it out, My Minimum value is "6.00000" or "6" & maximum value is
- "289.00000" or 289. This much difference we will be find out, between value and range.

- we should always go with "Data dispersion, or variability" & check should we scale our data or not?
    

In [11]:
data.describe().T

,count,mean,std,min,25%,50%,75%,max
fixed acidity,1599.0,8.319637,1.741096,4.60000,7.1000,7.90000,9.200000,15.90000
volatile acidity,1599.0,0.527821,0.179060,0.12000,0.3900,0.52000,0.640000,1.58000
citric acid,1599.0,0.270976,0.194801,0.00000,0.0900,0.26000,0.420000,1.00000
residual sugar,1599.0,2.538806,1.409928,0.90000,1.9000,2.20000,2.600000,15.50000
chlorides,1599.0,0.087467,0.047065,0.01200,0.0700,0.07900,0.090000,0.61100
free sulfur dioxide,1599.0,15.874922,10.460157,1.00000,7.0000,14.00000,21.000000,72.00000
total sulfur dioxide,1599.0,46.467792,32.895324,6.00000,22.0000,38.00000,62.000000,289.00000
density,1599.0,0.996747,0.001887,0.99007,0.9956,0.99675,0.997835,1.00369
pH,1599.0,3.311113,0.154386,2.74000,3.2100,3.31000,3.400000,4.01000
sulphates,1599.0,0.658149,0.169507,0.33000,0.5500,0.62000,0.730000,2.00000


# How we can perform the "Standara scala", So I can directly import from the "Standara scala" from this -

- URL : [ https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html#sklearn.svm.SVC ]
- Code : "from sklearn.preprocessing import StandardScaler"

- Let's Import this "Standara scala" and in "Scikit-learn" Class has been also given, "Scikit-learn standrad scala" 
- Below is the URL :  ["https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html"]

- So, here if we go & check PreProcessing module, of Scikit-learn, inside this Link we can find our many classes.
- Google it -> [sklearn.preprocessing ] 

- URL : https://scikit-learn.org/stable/api/sklearn.preprocessing.html

- Inside this "scikit-learn" we can find out many classes, 

- 1.Binarizer : Binarize data (set feature values to 0 or 1) according to a threshold. |
- URL : https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.Binarizer.html#sklearn.preprocessing.Binarizer
- 2.FunctionTransformer : Constructs a transformer from an arbitrary callable|
- URL : https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.FunctionTransformer.html#sklearn.preprocessing.FunctionTransformer
- 3.MinMaxScaler : Transform features by scaling each feature to a given range.|
- URL : https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler
- 4.MaxAbsScaler : Scale each feature by its maximum absolute value.|
- URL : https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MaxAbsScaler.html#sklearn.preprocessing.MaxAbsScaler

- Etc,.... there are other also, these are the few above one examples. Rest everything inside the URL.

- Lets say if we want to perfrom Encoding, there are many Classes, Going forward we are going to use many of them.
- These above URL inside there alot of other classes, check each of them, try to check with different different technique and while performing Encoding
- Our result might be different, we may get good accuracy interms of data, we might decrease also, try to understand the mathematical formula behind
- all these classes.

# "StandardScaler"

- Here in this URl : [ https://scikit-learn.org/stable/api/sklearn.preprocessing.html ] 
- there are many Classes, But over here we can see "StandardScaler" & 
- The formula or calculated as the "StandardScaler" is -> z = (x - u) / s. Based on this they will going to transformed the particular data.

# "robust_scale"

- 5.There is another class called the - "robust_scale" | Standardize a dataset along any axis.

- URL : [ https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.robust_scale.html#sklearn.preprocessing.robust_scale ]
- Inside this "robust_scale" they just changed the Formula while they are going to translate the data, here same thing Similar to "StandardScaler"
- is doing
- But here This document we can see they are using the different formula for Transforming a data.
- If we read this Document then we will better clearity atleast description part.
- If we open this URL and read this document, Over here will be see this line -> ("Scalte features using Statistics that are robust to outliers"
- Why this is Robust to Outliers?
- Because inside the formula, we are using the "Mean", if we check the "StandardScaler" formula then here this formula they are 
- using the "Mean". Here in this "StandardScaler" -> z = (x - u) <- "u" they are using the "Medium", incase of using the "Mean".
- "Standardization" required for the all the Columns, Yes we are going to perform for the all the column.

# Note :
 - we can check different different preProcessing technique, we can read the description of these all classes and implement it.

# "StandardScaler"

- As of now, I am going implement the "StandardScaler", into my Dataset.
- Over here, if we see "StandardScaler" is class, what I need to do is I just need to create the Object & I need to fit my data & I need to tranformed
- my particular data.
- Before applying the "StandardScaler" should we apply "train_test_split" or not? or we do a direct "StandardScaler"?

# Note : 

- So, we will going to perform the "train_test_split" & then we will going to perform the "StandardScaler" & then we will try to craete the model. 
- So what will be the Columns, I will be going to take the "Target" column is "X=data." and I am write drop.
- which feature we need to drop, inside my data. I will drop the "Quality" of the data. I will be geeting the data, without the "Quality" & I will 
- mention the "X-axis", the my "X-axis", would be the (1).
- So as I am able to drop the "Quality", if we check inside our "X" columns "X.head()", we are having the various features. Apart from the Qualities.

- Here I have dropped the "quality", I can't consider this, why because I am going to craete the Supervised model, 
- there I have to predict the "quality", I am just going to segregate the data a "X" & "Y".

In [12]:
X=data.drop("quality",axis=1)

In [13]:
X.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4


# Y Columns

- Now lets take the Y-Columns also, if we put the Quality and check and we can see this below is my "Y" features and above one is my "X" features.

In [14]:
y = data["quality"]

In [15]:
y.head()

0    5
1    5
2    5
3    6
4    5
Name: quality, dtype: int64

# Split Dataset

- Let's say If I want to split the dataset what I need do then, should I have to import the libaries ?

- We already discussed about the "train_test_split".
- So, we need to import the "train_test_split" after that, I need to create an object, for this "train_test_split".
- Here I am giving the "X", "Y", Test_size & Random_state.
- So, How many data we are taking for testing 33 % data we will take for the testing or we can say Validation of the Model.

- Lets execute this below code, as we can see that we are getting the "X_train" & "X_test" data. This way we can check and get to know that,
- how many rows and columns are there.

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [17]:
X_train.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol
548,12.4,0.350,0.49,2.6,0.079,27.0,69.0,0.99940,3.12,0.75,10.4
355,6.7,0.750,0.01,2.4,0.078,17.0,32.0,0.99550,3.55,0.61,12.8
1296,6.6,0.630,0.00,4.3,0.093,51.0,77.5,0.99558,3.20,0.45,9.5
209,11.0,0.300,0.58,2.1,0.054,7.0,19.0,0.99800,3.31,0.88,10.5
140,8.4,0.745,0.11,1.9,0.090,16.0,63.0,0.99650,3.19,0.82,9.6


In [18]:
y_train.head()

548     6
355     6
1296    5
209     7
140     5
Name: quality, dtype: int64

# Note :

- Now what we are going to do is that, we are "standardize" my data.I can performed my Standardzier with respect to the "X_train"
- If we are facing while importing the "Class" or "Model" so every time take help from the "Sklearn".

- So Google it -  StandardScaler -> URL : [ https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html ]

- If we open this document, there is already code is given to us, just copy few line of codes.

- 1.from sklearn.preprocessing import StandardScaler
- 2.scaler = StandardScaler()
- 
- What we have to we have to fit the data, before that we need to create the object of the "StandardScaler" class.

- Now, if we run the below code, 1st " StandardScaler" and we are able to create the object of the particular Class.
- Next, I am going to fit and transfer method, I am going to run it, all together also we can run it.

# Note : 

- In the document, just below the Code, there was the "Methods", For example -> ["fit_transform"], one method has been given.
- We can dirctly use this -> "fit_transform" for our dataset.
- So what we are going to do is, [ "Scaler.fit(data)" ] -> Scaler.fit & I am going to providing a data, lets provide the data ("X_train").
- "Standardization" only for the Train data, as of now what I am going to do is I am fitting my data a "X_train" & I am going to be transformed.
- Whatever mean & standard Deviation value I am going to get from the ("X_train").
- Same Value, I am going to use for my ( "X_test" ) data. 
- Again, I am not going to be fit, the pattern which I am going to read for my dataset, I am going to use for my "X_train".

# What I am doing over here is that ?

# Mean : print(scaler.mean_)

- I am fitting my data & I am Calculating "Mean" & "Standard Deviation", If we and check above URL & inside the Libary, Yes we can find out the 
- Mean, over here we can see -> "print(scaler.mean_)" we can find out the Mean, we can get the Mean of the data.
- If I will going to Print this -> "print(scaler.mean_)" <- Vaiable.
- I don't do the transform. what is the object of fitting the data. What I am going to do is that,
- I am going to calculate, "Mean" & "Standard Deviation".

- How many "Mean" value I am going to get from this particular object "scaler.fit(X_train)" ?
- If we run this -> [" print(scaler.mean_) " ], as we can see we are able to get the "Mean". With the each & every features.

# Transform : [ scaler.transform(X_train) ]

- Now, If Run this Transfer, what I can do I can run the Transfer of the data. I am going to perform the Transfer of the data.
- So, we will be able to find out that, I am able to transform entire feature.
- I need to pass it over here, ("X_train"). 
- As a outcome, we can see I am able to transform my data.

In [19]:
from sklearn.preprocessing import StandardScaler

In [20]:
scaler = StandardScaler()

In [21]:
scaler.fit(X_train)

StandardScaler()

In [22]:
print(scaler.mean_)

[ 8.30345472  0.53246499  0.26933707  2.54691877  0.08772736 15.91223156
 46.76330532  0.99677933  3.31453782  0.65881419 10.41521942]


In [23]:
scaler.transform(X_train)

array([[ 2.40069523, -1.03103722,  1.12742595, ..., -1.26096312,
         0.52726134, -0.01431863],
       [-0.93967131,  1.22920403, -1.32502245, ...,  1.52622836,
        -0.28225704,  2.24363201],
       [-0.99827424,  0.55113165, -1.37611513, ..., -0.74241587,
        -1.20742091, -0.86105011],
       ...,
       [-0.6466567 ,  0.49462562, -1.06955908, ...,  1.26695473,
        -0.68701624, -0.86105011],
       [-0.23643625, -1.87862768,  0.4121285 , ...,  0.03540501,
         0.81637505,  1.39690052],
       [-1.46709761, -1.3700734 , -0.04770558, ...,  0.48913386,
        -0.68701624,  2.90220094]])

# Public Question :  Why not "fit_transform"?

- We can use it, direct method also has been given, it will give me same thing, 
- lets copy and paste here from Document -> [ fit_transform(X, y=None, **fit_params) ] 
- It will give us same thing. What I can do is : "Scalter" object is there & "fit_transform" what I can pass the "X_train" & we can see similar result
- we can get, as we can see over here. 
- We are getting similar result or outcome just above one.
- Both are same whether, we write "Transformer" or "fit_transformer" 

In [24]:
scaler.fit_transform(X_train)

array([[ 2.40069523, -1.03103722,  1.12742595, ..., -1.26096312,
         0.52726134, -0.01431863],
       [-0.93967131,  1.22920403, -1.32502245, ...,  1.52622836,
        -0.28225704,  2.24363201],
       [-0.99827424,  0.55113165, -1.37611513, ..., -0.74241587,
        -1.20742091, -0.86105011],
       ...,
       [-0.6466567 ,  0.49462562, -1.06955908, ...,  1.26695473,
        -0.68701624, -0.86105011],
       [-0.23643625, -1.87862768,  0.4121285 , ...,  0.03540501,
         0.81637505,  1.39690052],
       [-1.46709761, -1.3700734 , -0.04770558, ...,  0.48913386,
        -0.68701624,  2.90220094]])

# Note : "X_train_tranform" | "Y"

- After doing this "Fit_tranform", I will keep it inside the variable name is -> [ "X_train_tranform" ] & I will keep inside this.
- SO, we have X train transform data & we have "Y" column :

- 1."X_train_tranform" &
- 2."Y"


In [25]:
X_train_tranform = scaler.fit_transform(X_train)

In [26]:
X_train_tranform

array([[ 2.40069523, -1.03103722,  1.12742595, ..., -1.26096312,
         0.52726134, -0.01431863],
       [-0.93967131,  1.22920403, -1.32502245, ...,  1.52622836,
        -0.28225704,  2.24363201],
       [-0.99827424,  0.55113165, -1.37611513, ..., -0.74241587,
        -1.20742091, -0.86105011],
       ...,
       [-0.6466567 ,  0.49462562, -1.06955908, ...,  1.26695473,
        -0.68701624, -0.86105011],
       [-0.23643625, -1.87862768,  0.4121285 , ...,  0.03540501,
         0.81637505,  1.39690052],
       [-1.46709761, -1.3700734 , -0.04770558, ...,  0.48913386,
        -0.68701624,  2.90220094]])

In [27]:
y

0       5
1       5
2       5
3       6
4       5
       ..
1594    5
1595    6
1596    6
1597    5
1598    6
Name: quality, Length: 1599, dtype: int64

# Let's go for Model Building.

- Before that, 
- Lets once create the Model then we will perform the Scalling.
- 1st Training and then testing.

# Note : Recap what How we will going to do this steps :

- Below is the steps we will going to follow :

- 1.Data ingestion - 1st we did the Data ingestion
- 2.EDA - Then we perform some short of EDA 
- 3.Pre-Processing - I can perform so many technique, I can check with different different technique, I can perform different different transform, 
- But as of now we have just did the "Standard Scaler" 
- 4.Model Building : Now we are ready for the "Model Building" rigth after Pre-Processing. Which "Model Building" we are going to use its a 
- "Support vector Classifier (SVC)"
- while we are doing the "Model Building" Here is the "data", or "data ingestion" after model Building, 
- then we need to perform the "Evalution" of the particular model.


- So, lets say this No.1 is my data and I perform the EDA & before this "Pre-Processing" I just Split my data, into the 2 parts : 
- "Train" & "Test".

- So, this "data" I can directly "pre-Process" & I can do the "Model Building". But what will be the use? How will I evaluate my model in that case.
- So, what we did, from No.1 ("data") here itself, I perform the "Split" & I did was I took the data for "Train" & "Test".
- Now for now, I am going to train my Model, with this Training data -> ("Train" & "Test").
- When I perform the "Evalution". How Will I do that, I will check with my "Test data", we can perform the "Cross validation" as well.

# We got Where I should used the "Test data"?

- I will validate or evaluated my data & I will evaluate my Model.

# 3.Pre-Processing 
- So, What I am going to do I am going to "Train" my model after the "Pre-Processing", 
- I am taking this "Train" data & I am going train my particular model

- Now for Training, I am going to use  - "Support vector Classifier" "SVC". 

# Note : "Cross Validation"

- "Cross Validation" we will going do tomorrow, not today.
- Inside that, we are just going to split the data based on a "Cross Validation". so, lets say we are taking( 5) so, 5 different different segment
- we are going to be split the data, 
- 1.one particular data we are going to be for the "Testing"
- 2.Anothe segment we are going to take for the "Training" Like this we are taking all over the data. that it.

# So lets start
- over here we are going to be Train_test_split.

#  Now we are going for the "Model Building" 

- Let continue with -"Support vector Classifier" "SVC"

# "Support vector Classifier" "SVC" : 

- Sklearn inside this URL : [ https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html#sklearn.svm.SVC ]
- Now we are going to use "Support vector Classifier" "SVC", For this ("SVC") They have given the "Class" inside the "Sklean" itself.
- I wil pick this piece of code from this URL code : 
- So, I will just import this particulat libary  :  (" from sklearn.svm import SVC " ) for implementing the "SVC" & then I will going to write
- Model=SVC() 

-  & what I am giving data, which data I am taking about. I am going to provide the "X_train_tranform" data which is my variable name and 
- inside this i have the data.

In [28]:
from sklearn.svm import SVC
model=SVC()

# Create a Model :

- 1st Lets create a model. Here I can "fit" or fitting my data, Lets keep ( "X & Y" ) both X_train & Y_train.
- If we excute this, as we can see I am able to create the Model.

In [47]:
model.fit(X_train_tranform,y_train)

SVC()

# Note :

- Now what we did was our Recap section, like what I did was that,
- we have craeted our model, we have train our model, and final outcome we can see that, this is my Final Model.

#  Now what next we have to do over here ?

- Now, we can check what accuracy and all we are getting with respect to data over here. 
- So, let me check with the ("train accuracy").
- I can perform the - ("model.score() this is the function inside here, what we can provide, I can give ("X_train_tranform") & ("y_train").
- As a outcome, we can see, I am able to find out the "Score" ? with respect to the data, I am able to get the score only (0.67) semething.
- My Model Training accuracy is 0.6778711484593838 or (0.67).
- I am taking about this Training accuracy -> inside "Split" there is - 1.("train")
- 1.Train
- 2.Test

In [30]:
model.score(X_train_tranform,y_train)

0.6778711484593838

- This SVC model which I have created, I will check with my "Test_data". How much accuracy and all I am getting.
- I am going to check with respect to my "test data" 
- So, model.predict, now Here what will I do is, I will write -> ("X_test").
- As we can see I am getting the value, when I am passing the "X_test", we are getting the Final outcome.
- As below outcome is the Predicted one, a predict data.

# Transformation :

- Here below I need to do a Transformation of the data, I have to "Pre-Process" ("X_test") data.
- So, what I need to do, scalling - "scaler" & I just need to call this function ( Tranform) -> I need to pass ("X_test)

In [31]:
scaler.transform(X_test)

array([[-3.53642095e-01,  1.55589436e-01, -9.67373729e-01, ...,
        -4.83142240e-01,  6.85666499e-03, -7.66968836e-01],
       [-2.95039173e-01, -1.83446751e-01, -5.07539654e-01, ...,
         4.89133857e-01, -1.03395269e+00, -8.61050113e-01],
       [ 1.40444556e+00,  7.77155778e-01, -2.52076279e-01, ...,
        -2.23868614e-01,  1.85718440e+00, -4.84725007e-01],
       ...,
       [-2.02456406e-03, -1.25706134e+00,  6.16499196e-01, ...,
        -2.94133945e-02,  6.42906824e-01,  1.96138818e+00],
       [-6.06274859e-02,  4.50655383e+00, -1.37611513e+00, ...,
         1.39659155e+00, -9.76129945e-01,  4.56087756e-01],
       [ 4.66798811e-01,  7.20649747e-01, -6.09725004e-01, ...,
        -2.23868614e-01, -6.87016236e-01, -7.66968836e-01]])

- Now here what I am doing is, I am ("X_test_transform"). This data I am providing to my "predict" function. 
- Lets see what, I am geeting now.
- As a outcome we can see I am geeting all the value different.

In [45]:
X_test_tranform = scaler.transform(X_test)

In [46]:
model.predict(X_test_tranform)

array([5, 5, 6, 5, 6, 5, 5, 5, 6, 6, 6, 5, 6, 5, 5, 7, 5, 6, 7, 5, 5, 5,
       6, 6, 5, 5, 6, 5, 5, 6, 5, 5, 6, 5, 6, 5, 6, 6, 5, 6, 5, 5, 6, 5,
       6, 6, 6, 6, 5, 6, 5, 5, 6, 7, 5, 5, 6, 5, 6, 5, 6, 6, 5, 5, 7, 5,
       6, 5, 7, 5, 6, 5, 6, 6, 6, 5, 7, 5, 6, 7, 5, 7, 5, 5, 6, 6, 5, 6,
       6, 5, 6, 5, 5, 6, 5, 6, 5, 6, 5, 5, 5, 5, 6, 6, 6, 6, 6, 5, 6, 5,
       6, 5, 6, 5, 6, 6, 6, 5, 5, 6, 6, 6, 6, 5, 5, 5, 6, 6, 5, 6, 6, 5,
       5, 6, 6, 5, 5, 5, 5, 6, 6, 6, 6, 5, 6, 5, 6, 5, 6, 5, 6, 6, 5, 6,
       6, 6, 5, 6, 5, 6, 7, 6, 6, 5, 5, 6, 5, 5, 5, 5, 5, 5, 6, 5, 7, 6,
       6, 5, 5, 5, 5, 7, 5, 7, 5, 6, 6, 6, 7, 5, 6, 6, 5, 6, 6, 5, 5, 5,
       6, 6, 5, 5, 5, 5, 7, 6, 5, 5, 6, 6, 7, 5, 6, 6, 6, 6, 6, 5, 6, 5,
       5, 6, 6, 6, 5, 5, 5, 7, 5, 5, 5, 5, 6, 6, 5, 6, 5, 6, 6, 5, 5, 5,
       6, 6, 5, 6, 6, 5, 6, 5, 6, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 5, 7,
       6, 7, 6, 5, 6, 6, 5, 6, 5, 5, 5, 5, 6, 6, 6, 5, 7, 5, 5, 5, 5, 6,
       5, 6, 5, 6, 5, 7, 6, 5, 5, 6, 5, 6, 6, 7, 5,

In [37]:
y_predict = model.predict(X_test_tranform)

# Note : 

- Now can we compaire with my Actual data Vs Predicted data.

- Below is my "Actual Value" ("y_test")

In [34]:
y_test

803     6
124     5
350     6
682     5
1326    6
       ..
813     4
377     7
898     7
126     5
819     5
Name: quality, Length: 528, dtype: int64

# Public Question : Why not ("fit.transform")

- Because "fit" over here we can see, in top we already did check there, this line code check above -> [ "scaler.fit(X_train)" ].

- Here - [ "scaler.fit(X_train)" ] - I am doing fitting. & 
- I am able to read this patten in outcome, from my dataset, we can check this line code : [" print.scaler.mean_ " ) 
- means -> [" print.scaler.mean_ ") here outcome :

- [ 8.30345472  0.53246499  0.26933707  2.54691877  0.08772736 15.91223156
-  46.76330532  0.99677933  3.31453782  0.65881419 10.41521942]

- Means this is the mean value -> 8.30345472 | 0.53246499 | 0.26933707 etc, from the particular feature. 
- So, the same "Mean" value, I am going to use for my "Test data". This one -> ( "X_test_tranform" ).

# Why I am doing this?

- Because I want to preserved the same pattern, from the train data -> ( "X_train_tranform" ). 
- Beacuse any how I am going to tarin my model on my "train" data, to avoid the "Data leakage" 

# Next,

- Now after prediction what I need to "Should I check with the accuracy score"

# Next: 

- So, from Sklearn.mertics import accuracy_score.
- accuracy_score() <- Inside here what I need to provide, this -> ( " X_test_tranform " ) value, see above code, we will get.
- I will keep this value this line of code -> ( "model.predict(X_test_tranform) " ) inside this variable -> [ "y_predict" ].
- Now, Below 2nd line code, ->  I will provide the (" y_test & y_predict ").
- As a outcome, how much accuracy we are getting we can see over here, 0.5984848484848485 or ( 59 % ) accuracy. 
- Here, we are checking the Testing accuracy.

# Note : 

- ("model.score") -> is basically for the "Training accuracy" & Here, below we are checking the Testing accuracy.
- So, How I can incrase the Accuracy?
- Can I check with the "Logistic Regression" also, How I say this is the best accuracy? 
- Even those "Logistic Regression" will be useful for the "Binary Classification" but one vs rest technique, & one vs one it will be there.
- So, we can solve the "Classification problem also.
- We can build that ""Logistic Regression" also. Just Google it -> [ " sklearn Logistic Regression "].
- URL : [ https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html ]. If we open this and inside this document
- Here inside this "Logistic Regression", we are having one parameters.
- This line check -> [ multi_class{‘auto’, ‘ovr’, ‘multinomial’}, default=’auto’ ] <- Here we are having one vs rest & one vs one.

- So, what I am going do, I am going to create an "Model" a "Logistic Regression Modle".
- If we check inside this document there is code, copy this line code ->  [ " from sklearn.linear_model import LogisticRegression " ]
- I am going to import it. Now we can check accuracy and compare it. Because so far we only learn the "Logistic Regression" model, this is why we are
- discussing, we can compare the accuracy.
- So to incrase the accuracy of this line code -> ( accuracy_score(y_test,y_predict) ) outcome -> ( 0.5984848484848485 or 59 % ) what we can do.
- we will discuss later.

In [36]:
from sklearn.metrics import accuracy_score

In [38]:
accuracy_score(y_test,y_predict)

0.5984848484848485

- Here what we are doing, just import the "Logistic Regression" 

#  2nd line code :
- I am going to write the Model No.2, & let me write "Logistic Regression", this is my model & which parameter, i need to pass.
- For the "Multi Class classification" -> this one -> [ {‘auto’, ‘ovr’, ‘multinomial’}, default=’auto’ ] <- Auto means one vs rest | multinomial.
- So, automatically, it will take this one (‘auto’,) & Default one will be ‘auto’.
- ‘auto’ means "one vs rest", non it to define, it will automatically take One Vs rest for the "Multi class Classification" problem statement.

# Hyper parameter
- If we look into the "Hyper Parameter" its a same thing. ‘auto’ will be there & ‘auto’ means one vs rest. Like this we can navigate.

# 3rd Line code :  

- Now If I am going to create a model, what I need to do,  model2 and I am going to fit my the data, what data. I am going to fit.
- [ " X_train_transform " & y_train ] <- Now if I will execute, As a outcome, we can see this is the model, which I am going to fit.

In [42]:
from sklearn.linear_model import LogisticRegression

In [43]:
model2 = LogisticRegression()

In [52]:
model2.fit(X_train_tranform,y_train)

LogisticRegression()

# Next :

- Can I do the predict with respect to my data?

- model2.predict <- this is the function, which I need to call, & here I need to pass the "Test data" which is ("X_test") 
- As a outcome, we can see, we are getting this much value.

In [49]:
model2.predict(X_test)

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2732: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(


array([5, 5, 5, 5, 5, 5, 5, 7, 5, 5, 7, 7, 5, 5, 5, 7, 5, 5, 7, 5, 7, 5,
       7, 7, 7, 5, 5, 5, 5, 7, 5, 5, 7, 7, 5, 5, 7, 7, 7, 7, 7, 5, 7, 5,
       7, 7, 7, 7, 5, 7, 5, 5, 7, 5, 5, 5, 7, 5, 5, 5, 5, 7, 5, 5, 7, 5,
       7, 5, 7, 5, 5, 5, 5, 5, 7, 5, 5, 5, 5, 5, 5, 7, 5, 7, 5, 5, 5, 7,
       5, 7, 5, 5, 7, 7, 5, 7, 5, 7, 7, 5, 7, 5, 5, 7, 5, 5, 7, 5, 5, 5,
       5, 5, 5, 5, 7, 7, 7, 5, 5, 5, 5, 5, 5, 5, 7, 5, 5, 5, 5, 5, 5, 7,
       5, 7, 7, 5, 5, 5, 5, 5, 7, 5, 5, 7, 7, 5, 7, 7, 5, 5, 5, 7, 5, 5,
       5, 5, 5, 5, 5, 7, 7, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 7, 7, 5, 5, 5,
       5, 5, 5, 5, 7, 5, 5, 7, 7, 5, 7, 5, 7, 5, 5, 5, 5, 5, 5, 5, 5, 5,
       5, 5, 7, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 7, 5, 7, 7, 5, 5, 5, 5,
       5, 5, 5, 5, 5, 5, 5, 7, 5, 5, 5, 5, 7, 7, 7, 5, 7, 5, 7, 5, 7, 5,
       5, 7, 5, 5, 7, 5, 7, 5, 5, 7, 7, 5, 7, 5, 5, 7, 7, 5, 5, 7, 5, 7,
       7, 7, 5, 5, 5, 5, 5, 5, 5, 5, 7, 5, 5, 5, 5, 5, 7, 7, 5, 5, 5, 5,
       5, 5, 5, 5, 5, 7, 7, 5, 5, 5, 5, 7, 5, 7, 5,

# Next :

- I can keep this inside the 2nd variable [ "y_predict2" ] <- inside this vaiable I will keep this things.
-  Now, I am going to check with the Accuracy_score.  Copy the code from above line this one -> [ accuracy_score(y_test,y_predict) ]
- So, what will be my actual data, my actual data will be "y_test" & my predicted data will be -> "y_predict2".
- Now, what will be the outcome, will I able to increase the accuracy or decrease the accuracy?
- As a outcome, we can see I am getting 0.571969696969697 or ( 57 % ), as we can see as compare to previous one I am getting less accuracy which is 57%.
- which mean its decesased. 

- # Interviewer : 

- So, like this we can check with the "Multi Model" before applying the model.
- Mathematical concept is also important because, interviewer will ask us.
- 1.why we are getting this accuracy ? 
- 2.can you explain in mathematical way etc.
- 3.Can you explain the "Loss Function" ? 
- 4.can you explain the Training methology for the SVM? 
- we can perform different different Feature Engineering technique to improve the accuracy.

# "Bulk" value prediction :

- So, this all we have did was the "Bulk" value prediction.

In [54]:
y_predict2 = model2.predict(X_test_tranform)

In [55]:
accuracy_score(y_test,y_predict2)

0.571969696969697

# Single Value Prediction ;

- Lets try to predict the "Single_Value_Prediction".
- Before that lets me call the -> [ "X_test_tranform" ] <- what is this, this is a "Array". Means after the "Standardization" we got the Array.
- we got the "Array" based on my "Test data"
- This is my data, different different rows has been given, I splited my data.
- SO, I will going to fetch my 1st "row" -> [ X_test_tranform[0] ]
- I am passing this 1st Row to predict function.

In [57]:
X_test_tranform

array([[-3.53642095e-01,  1.55589436e-01, -9.67373729e-01, ...,
        -4.83142240e-01,  6.85666499e-03, -7.66968836e-01],
       [-2.95039173e-01, -1.83446751e-01, -5.07539654e-01, ...,
         4.89133857e-01, -1.03395269e+00, -8.61050113e-01],
       [ 1.40444556e+00,  7.77155778e-01, -2.52076279e-01, ...,
        -2.23868614e-01,  1.85718440e+00, -4.84725007e-01],
       ...,
       [-2.02456406e-03, -1.25706134e+00,  6.16499196e-01, ...,
        -2.94133945e-02,  6.42906824e-01,  1.96138818e+00],
       [-6.06274859e-02,  4.50655383e+00, -1.37611513e+00, ...,
         1.39659155e+00, -9.76129945e-01,  4.56087756e-01],
       [ 4.66798811e-01,  7.20649747e-01, -6.09725004e-01, ...,
        -2.23868614e-01, -6.87016236e-01, -7.66968836e-01]])

In [58]:
X_test_tranform[0]

array([-0.3536421 ,  0.15558944, -0.96737373, -0.03334372,  0.55556956,
       -0.18596079, -0.02314512,  0.1740298 , -0.48314224,  0.00685666,
       -0.76696884])

# Note :

- So, whatever outcome, I got in my 1st row [0], what I will going do is, I am going to copy this outcome rows and paste it here and pass it.
- So the "model.predict" & these all the rows, I am going to pass.

In [59]:
model.predict([[-0.3536421 ,  0.15558944, -0.96737373, -0.03334372,  0.55556956,
       -0.18596079, -0.02314512,  0.1740298 , -0.48314224,  0.00685666,
       -0.76696884]])

array([5])

# Note :  "Single Value prediction"

- [ pd.read_csv("https://raw.githubusercontent.com/aniruddhachoudhury/Red-Wine-Quality/master/winequality-red.csv")  ]

- Based on these paticular features -[ Fixed acidity | Volatile acidity	| citric acid |	residual sugar | chlorides | free sulfur dioxide |
- total sulfur dioxide | density | pH |	sulphates |	alcohol ]<- I wa going to predict the [ "quality" ] of the wine. we can see in our data.
    
- What we are doing over here, we have trained the model, & we are passing this "Independent features" where I am passing.

-  [ model.predict([[-0.3536421 ,  0.15558944, -0.96737373, -0.03334372,  0.55556956,
-    -0.18596079, -0.02314512,  0.1740298 , -0.48314224,  0.00685666,
-      -0.76696884]])

- This outcome is [5] this is the Quality of wine, means we are passing this Row of value this one -> [ 0.3536421 ,  0.15558944, -0.96737373,etc]

- Means we are passing this "row value" not columns , as we ca see in our 1st row.
- 0 | 7.4 | 0.70 |0.00 | 1.9 | 0.076| 11.0 | 34.0 | 0.9978 | 3.51  | 0.56 | 9.4| 5 |
- Based on these all features, this is giving me that Quality of the data. 
- This is called the "Single Value prediction" 

# Note : "Gridsearch CV" | Class work HomeWork

- Lets complete this "Kernal" part & "ROC AUC Score" then we will jump to the -> "Gridsearch CV".
- Then we will try to increase the accuracy for this particular dataset -> [ accuracy_score(y_test,y_predict) ] 

- Let's say if we want to check or play with the accuracy, if we want to play with the "Hyper Parameter" then go to this document in Sklearn.
- Google it -> "sklearn greatsearch CV" & and go and check this class.

-> URL : [ https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html ] 

- So, from this document we can import the class and "Gridsearch CV" class and we can make the Object. We can check with the differnt different
- "Hyper parameter" also.
- As we can see by default, this line code I picked my document -> [ svc = svm.SVC() ] <- SVC example has been given over here.

# Note : SkLearn 

- Sklearn is nothing python code only. Someone has written class for us to use it dircet by importing or calling the module.
- SkLearn if we search Sklearn Github, and open it, we can see this is end to end project. Its a same project someone has written, Classes for us.
- If we open this SKlearn and check this in Github, open any let's say I open this -> "Logistic.py" new tab we can see over here.
- Someone has written same mathematic concept in Python, he has implemented same mathematics, in python itself, whatever we can see.
- If we open the document, then same this was that, so dircetly we are calling this function and classes in our code.
- If we import or load anything from the Sklearn it will dirctly come here.
- We can build our own package and Libary. we have discussed this previously python lecture where we have discussed about the "Module and Packages"
    
- "Sklearn" is nothing python code only someone has written, class for all of us.

# Note : Homework -> SVR ("Support Vector Regression")

- To implement the SVR by ourself? Below is the Dataset for Practical.

- URL Dataset : pd.read_csv("https://raw.githubusercontent.com/srinivasav22/Graduate-Admission-Prediction/master/Admission_Predict_Ver1.1.csv")

- This is the famous dataset, we have previously also implemented, we just have to predict the "Admission chances" 

- The columns name ("Chance of the admission" ) will be my target column. This will be my "dependent" columns.

- Rest coumns -> [ "Serial No."	"GRE Score"	"TOEFL Score" "University Rating" ] <- These are my "Independent" columns.

In [56]:
pd.read_csv("https://raw.githubusercontent.com/srinivasav22/Graduate-Admission-Prediction/master/Admission_Predict_Ver1.1.csv")

,Serial No.,GRE Score,TOEFL Score,University Rating,SOP,LOR,CGPA,Research,Chance of Admit
0,1,337,118,4,4.5,4.5,9.65,1,0.92
1,2,324,107,4,4.0,4.5,8.87,1,0.76
2,3,316,104,3,3.0,3.5,8.00,1,0.72
3,4,322,110,3,3.5,2.5,8.67,1,0.80
4,5,314,103,2,2.0,3.0,8.21,0,0.65
...,...,...,...,...,...,...,...,...,...
495,496,332,108,5,4.5,4.0,9.02,1,0.87
496,497,337,117,5,5.0,5.0,9.87,1,0.96
497,498,330,120,5,4.5,5.0,9.56,1,0.93
498,499,312,103,4,4.0,5.0,8.43,0,0.73


# TASK : Above is the Dataset.

# Questions : 

1.YOU HAVE TO INCREASE THE ACCURACY OF THE SVC MODEL(WINEQUALITY DATASET)

2.HYPERPARAMETER TUNING(GRIDSEARCH cv https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)

3.YOU HVAE TO IMPLEMENT SVR(ADDIMISSOIN_PREDICTION)

- https://raw.githubusercontent.com/srinivasav22/Graduate-Admission-Prediction/master/Admission_Predict_Ver1.1.csv